rust/headless-client/src/windows.rs

```diff
@@ -29,3 +230,79 @@ pub(crate) fn default_token_path() -> std::path::PathBuf {
 pub(crate) fn notify_service_controller() -> Result<()> {
     Ok(())
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use firezone_bin_shared::TunDeviceManager;
+    use ip_network::Ipv4Network;
+    use socket_factory::DatagramOut;
+    use std::borrow::Cow;
+    use std::net::{Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4};
+    use std::time::Duration;
+
+    #[test]
+    fn best_route_ip4_does_not_fail() {
+        get_best_route_excluding_interface("8.8.8.8".parse().unwrap(), "Firezone");
+    }
+
+    #[test]
+    fn best_route_ip6_does_not_fail() {
+        get_best_route_excluding_interface("2404:6800:4006:811::200e".parse().unwrap(), "Firezone");
+    }
+
+    // Starts up a WinTUN device, adds a "full-route" (`0.0.0.0/0`) and checks if we can still send packets to IPs outside of our tunnel.
+    #[tokio::test]
+    #[ignore = "Needs admin & Internet"]
+    async fn no_packet_loops() {
+        let ipv4 = Ipv4Addr::from([100, 90, 215, 97]);
+        let ipv6 = Ipv6Addr::from([0xfd00, 0x2021, 0x1111, 0x0, 0x0, 0x0, 0x0016, 0x588f]);
+
+        let mut device_manager = TunDeviceManager::new().unwrap();
+        let _tun = device_manager.make_tun().unwrap();
+        device_manager.set_ips(ipv4, ipv6).await.unwrap();
+
+        // Configure `0.0.0.0/0` route.
+        device_manager
+            .set_routes(
+                vec![Ipv4Network::new(Ipv4Addr::UNSPECIFIED, 0).unwrap()],
+                vec![],
+            )
+            .await
+            .unwrap();
+
+        // Make a socket.
+        let mut socket =
+            udp_socket_factory(&SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::UNSPECIFIED, 0)))
+                .unwrap();
+
+        // Send a STUN request.
+        socket
+            .send(DatagramOut {
+                src: None,
+                dst: SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(141, 101, 90, 0), 3478)), // stun.cloudflare.com,
+                packet: Cow::Borrowed(&hex_literal::hex!(
+                    "000100002112A4420123456789abcdef01234567"
+                )),
+            })
+            .unwrap();
+
+        // First send seems to always result as would block
+        std::future::poll_fn(|cx| socket.poll_flush(cx))
+            .await
+            .unwrap();
```

Comment:
@thomaseizinger flushing the socket seems to be what the test was missing, works with this! :)

Reply:
Awesome! Thanks for debugging!

Reply:
Did you confirm that it fails without our "fix" of setting the source address?

Reply:
It'd be nice to expand this test to the other platforms in a follow up

Reply:
This should work as is on Linux I think?

Reply:
yeah, I think so